cleanlab/internal/neighbor/knn_graph.py

```diff
@@ -0,0 +1,335 @@
+    indices :
+        The indices of the k nearest neighbors for each point.
+    enable_warning :
+        Whether to enable warning messages if any row underestimates the number of exact duplicates.
+
+    Returns
+    -------
+    corrected_distances :
+        The corrected distances between each point and its k nearest neighbors. Exact duplicates (based on the feature array) are ensured to have zero mutual distance.
+    corrected_indices :
+        The corrected indices of the k nearest neighbors for each point. Exact duplicates are ensured to be included in the k nearest neighbors, unless the number of exact duplicates exceeds k.
+
+    Raises
+    ------
+    UserWarning :
+        A warning may be raised if there were some slots available for an exact duplicate that were missed.
+        This may happen if the number of exact duplicates in the existing knn graph is lower than k,
+        but the set of exact duplicates is larger than what was included in the knn graph.
+        This warning may be disabled by setting enable_warning=False.
+
+
+    Example
+    -------
+    >>> import numpy as np
+    >>> X = np.array(
+    ...     [
+    ...         [0, 0],
+    ...         [0, 0], # Exact duplicate of the previous point
+    ...         [1, 1], # The distances between this point and the others is sqrt(2) (equally distant from both)
+    ...     ]
+    ... )
+    >>> distances = np.array(  # Distance to the 1-NN of each point
+    ...     [
+    ...         [np.sqrt(2)],  # Should be [0]
+    ...         [1e-16],       # Should be [0]
+    ...         [np.sqrt(2)],
+    ...     ]
+    ... )
+    >>> indices = np.array(  # Index of the 1-NN of each point
+    ...     [
+    ...         [2],  # Should be [1]
+    ...         [0],
+    ...         [1],  # Might be [0] or [1]
+    ...     ]
+    ... )
+    >>> corrected_distances, corrected_indices = correct_knn_distances_and_indices(X, distances, indices)
+    >>> corrected_distances
+    array([[0.], [0.], [1.41421356]])
+    >>> corrected_indices
+    array([[1], [0], [0]])
+
+
+    Clearly, the first point misses its exact duplicate in the knn graph. To raise a warning for such cases, set enable_warning=True.
+
+    >>> corrected_distances, corrected_indices = correct_knn_distances_and_indices(X, distances, indices, enable_warning=True)
+    UserWarning: There were some slots available for an exact duplicate that were missed.
+    """
+
+    # Number of neighbors
+    k = distances.shape[1]
+
+    # Prepare the output arrays
+    corrected_distances = np.zeros_like(distances)
+    corrected_indices = np.zeros_like(indices, dtype=int)
+
+    # Use np.unique to catch inverse indices of all unique feature sets
+    unique_inverse = np.unique(features, return_inverse=True, axis=0)[1]
+
+    # Map each unique feature set to its indices across the dataset
+    feature_map = {u: np.where(unique_inverse == u)[0] for u in set(unique_inverse)}
```

Comment:
could we name this variable more explicitly, like: `exact_duplicate_map` ?

Reply:
Switched from a map to a list of `exact_duplicate_sets`.